SXDEDPCXZIC-31_DATAVIC-430 / updates to DELWP harvester

alexmorev opened the pull request:
Need additional changes to ckan.ini:
## ckanext-datavic-harvester
ckanext.datavic_harvester.geoserver_dns.wms = https://geoserver-uat.maps.vic.gov.au/geoserver/ows?service=WMS&request=getCapabilities
ckanext.datavic_harvester.geoserver_dns.wfs = https://geoserver-uat.maps.vic.gov.au/geoserver/ows?service=WFS&request=getCapabilities
ckanext.datavic_harvester.resource.wms_url = https://geoserver-uat.maps.vic.gov.au/geoserver/wms?service=wms&request=getmap&format=image%%2Fpng8&transparent=true&layers={}&width=512&height=512&crs=epsg%%3A3857&bbox=16114148.554967716%%2C-4456584.4971389165%%2C16119040.524777967%%2C-4451692.527328665
ckanext.datavic_harvester.resource.wfs_url = https://geoserver-uat.maps.vic.gov.au/geoserver/wfs?request=GetCapabilities&service=WFS

alexmorev committed:
SXDEDPCXZIC-31_DATAVIC-430 / updates to DELWP harvester

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -6,13 +6,15 @@
 from ckanext.harvest.model import HarvestObject, HarvestObjectExtra
 from hashlib import sha1
 from ckan.plugins import toolkit as toolkit
+from ckan.common import config
```
toolkit.config

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -447,6 +449,48 @@ def _get_package_dict(self, harvest_object, context):
+                if attribution:
+                    wms_res['attribution'] = attribution
+                resources.append(wms_res)
+        except Exception as e:
```
what kind of error are we expecting here?

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -447,6 +449,48 @@ def _get_package_dict(self, harvest_object, context):
+                if attribution:
+                    wfs_res['attribution'] = attribution
+                resources.append(wfs_res)
+        except Exception as e:
```
what kind of error are we expecting here?

mutantsan replied:
As I said to you a few days ago, you have to include in the try/except only code, that could throw an exception. You must know what every line of your code is doing and what could go wrong.

mutantsan replied:
+ Your code is self repeating. Read about `DRY` principle and elaborate an approach, when you are reusing code, instead of copying it. Just find the differences and similarities first.

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -447,6 +449,48 @@ def _get_package_dict(self, harvest_object, context):
 
+        # Verify and get info for WMS resource       
+        try:
+            wms_doc = requests.get(config.get("ckanext.datavic_harvester.geoserver_dns.wms"))
```
why did you add it to the CKAN config, instead of harvest source config? What if we want multiple Delwp harvesters using different geoserver_dns?

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -447,6 +449,48 @@ def _get_package_dict(self, harvest_object, context):
+            wms_doc = requests.get(config.get("ckanext.datavic_harvester.geoserver_dns.wms"))
+            wms_soup= BeautifulSoup(wms_doc.content,"lxml-xml")
+            wms_obj = wms_soup.find_all("Layer", {"queryable": 1})
+            wms_result = [x for x in wms_obj if uuid in str(x)]
```
what is x?

mutantsan replied:
What the hell is `str(x)`? Are you converting the whole Layer data to string and looking for inclusion?

mutantsan replied:
If you want to compare the ID's, you must be precise. 
`layer.find("KeywordList").findAll("Keyword", string=f"MetadataID={uuid}")`

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -447,6 +449,48 @@ def _get_package_dict(self, harvest_object, context):
+        try:
+            wms_doc = requests.get(config.get("ckanext.datavic_harvester.geoserver_dns.wms"))
+            wms_soup= BeautifulSoup(wms_doc.content,"lxml-xml")
+            wms_obj = wms_soup.find_all("Layer", {"queryable": 1})
```
As I understand, `wms_obj` is a list of Layers. It's a good idea to work on namings and use typing to write a better code

alexmorev committed:
SXDEDPCXZIC-31_DATAVIC-430 / optimize the changes

alexmorev commented:
Fixed all bugs and implemented the mentioned changes.
Need the following additional changes to harvest source config:
"add_resources": 
            {
                "wms":
                    {
                        "full_url": "https://geoserver-uat.maps.vic.gov.au/geoserver/ows?service=WMS&request=getCapabilities",
                        "resource_url": "https://geoserver-uat.maps.vic.gov.au/geoserver/wms?service=wms&request=getmap&format=image%2Fpng8&transparent=true&layers={layername}&width=512&height=512&crs=epsg%3A3857&bbox=16114148.554967716%2C-4456584.4971389165%2C16119040.524777967%2C-4451692.527328665"
                    },
                "wfs":
                    {
                        "full_url": "https://geoserver-uat.maps.vic.gov.au/geoserver/ows?service=WFS&request=getCapabilities",
                        "resource_url": "https://geoserver-uat.maps.vic.gov.au/geoserver/wfs?request=GetCapabilities&service=WFS"
                    }
            }

alexmorev committed:
SXDEDPCXZIC-31_DATAVIC-430 / fix exeptions bugs

alexmorev commented:
Fixed exceptions bugs and harvest source config requirements for WMS/WFS resources.
Need the following additional changes to harvest source config:
"geoserver_dns": "https://geoserver-uat.maps.vic.gov.au"

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -441,12 +433,53 @@ def _get_package_dict(self, harvest_object, context):
-
+                
+        # Generate additional WMS/WFS resources     
+        if self.config['geoserver_dns']:
```
KeyError

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -441,12 +433,53 @@ def _get_package_dict(self, harvest_object, context):
+            dict_geoserver_urls = {
+                'WMS': [
+                    geoserver_dns + '/geoserver/ows?service=WMS&request=getCapabilities',
+                    'https://geoserver-uat.maps.vic.gov.au/geoserver/wms?service=wms&request=getmap&format=image%2Fpng8&transparent=true&layers={layername}&width=512&height=512&crs=epsg%3A3857&bbox=16114148.554967716%2C-4456584.4971389165%2C16119040.524777967%2C-4451692.527328665'
```
Hardcoded resource URL

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -441,12 +433,53 @@ def _get_package_dict(self, harvest_object, context):
+                ],
+                'WFS': [
+                    geoserver_dns + '/geoserver/ows?service=WFS&request=getCapabilities',
+                    'https://geoserver-uat.maps.vic.gov.au/geoserver/wfs?request=GetCapabilities&service=WFS'
```
Hardcoded resource URL

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -441,12 +433,53 @@ def _get_package_dict(self, harvest_object, context):
+                ]
+            }
+            
+            def _get_content_with_uuid(geoserver_url):
```
Why inner function?

mutantsan replied:
https://realpython.com/inner-functions-what-are-they-good-for/#:~:text=Inner%20functions%2C%20also%20known%20as,closure%20factories%20and%20decorator%20functions.

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -441,12 +433,53 @@ def _get_package_dict(self, harvest_object, context):
+                    return None
+                try:
+                    geoserver_content= BeautifulSoup(res.content,"lxml-xml")
+                except NameError as e:
```
Why NameError here?

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -441,12 +433,53 @@ def _get_package_dict(self, harvest_object, context):
+                ]
+            }
+            
+            def _get_content_with_uuid(geoserver_url):
```
You need a layer related to a dataset UUID. But you are returning something like this:
`<Keyword>MetadataID=bc822a9c-3766-57ac-a034-bcad3fb66d86</Keyword>`

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -441,12 +433,53 @@ def _get_package_dict(self, harvest_object, context):
+                except NameError as e:
+                    log.error(e)
+                    return None
+                layer_data_with_uuid = geoserver_content.find("Keyword", string=f"MetadataID={uuid}")
```
What's the point of creating a variable instead of just returning this?

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -441,12 +433,53 @@ def _get_package_dict(self, harvest_object, context):
+                
+            def generate_geo_resource(layer_data_with_uuid, format, resource_url):
+                resource_data = {
+                    "name": layer_data_with_uuid.find_previous("Title").text.upper() + ' ' + format.upper(),
```
Why `upper` everywhere, if it's hardcoded in uppercase?

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -441,12 +433,53 @@ def _get_package_dict(self, harvest_object, context):
+                return resource_data
+            
+            for format, urls in dict_geoserver_urls.items():
+                layer_data = _get_content_with_uuid(urls[0])
```
Magic numbers 0 and 1.

mutantsan replied:
https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad

mutantsan replied:
Also, I suggest you read this. https://linkdigital.jira.com/wiki/spaces/LBFAC/pages/2152497307/Python+typing
Start using typing ASAP

mutantsan commented on ckanext/datavic_harvester/harvesters/delwp.py:
```diff
@@ -441,12 +433,53 @@ def _get_package_dict(self, harvest_object, context):
+                }
+                return resource_data
+            
+            for format, urls in dict_geoserver_urls.items():
```
Don't mess with the `format` var name too much, because we have `.format` method in python. It could be confusing. Use something like `res_format`, `fmt`, `resource_format`, `res_fmt`, `_format`

alexmorev committed:
SXDEDPCXZIC-31_DATAVIC-430 / fixed new bugs

alexmorev committed:
SXDEDPCXZIC-31_DATAVIC-430 / fix using inner func

joelwigley committed:
Merge pull request #26 from agmorev/SXDEDPCXZIC-31_DATAVIC-430

SXDEDPCXZIC-31_DATAVIC-430 / updates to DELWP harvester

muhammed-ajmal committed:
SXDEDPCXZIC-44_DATAVIC-493/cleaned resource name

muhammed-ajmal committed:
SXDEDPCXZIC-44_DATAVIC-493/cleaned resource name

muhammed-ajmal committed:
Merge branch 'develop' into SXDEDPCXZIC-44_DATAVIC-493

Engerrs committed:
Merge pull request #29 from muhammed-ajmal/SXDEDPCXZIC-44_DATAVIC-493

SXDEDPCXZIC-44_DATAVIC-493/cleaned resource name [develop]

mutantsan committed:
SXDEDPCXZIC-90_DATAVIC-470 / black project

mutantsan committed:
SXDEDPCXZIC-90_DATAVIC-470 / Remove redundant directories and files

mutantsan committed:
SXDEDPCXZIC-90_DATAVIC-470 / move harvesters

mutantsan committed:
SXDEDPCXZIC-90_DATAVIC-470 / refine imports and related code

mutantsan committed:
SXDEDPCXZIC-90_DATAVIC-470 / refactor helpers

mutantsan committed:
SXDEDPCXZIC-90_DATAVIC-470 / replace % and format string interpolation

smotornyuk committed:
SXDEDPCXZIC-89_DATAVIC-517 / prepare ODS description using markdownify

smotornyuk committed:
Merge pull request #30 from dpc-sdp/SXDEDPCXZIC-89_DATAVIC-517

Sxdedpcxzic 89 datavic 517

mutantsan committed:
SXDEDPCXZIC-90_DATAVIC-470 / commit last changes to datavic_ckan and metashare harvesters

mutantsan committed:
SXDEDPCXZIC-90_DATAVIC-470 / remove datavic_ckan and metashare harvesters, client is not using it

mutantsan committed:
SXDEDPCXZIC-90_DATAVIC-470 / refactor delwp harvester (wip)

mutantsan committed:
SXDEDPCXZIC-90_DATAVIC-470 / finish refactor delwp harvest

mutantsan committed:
SXDEDPCXZIC-90_DATAVIC-470 / finish refactor dcat_json and small fixes for delwp

mutantsan committed:
Merge remote-tracking branch 'origin/develop' into SXDEDPCXZIC-90_DATAVIC-470

mutantsan committed:
SXDEDPCXZIC-109 / writing tests for base and delwp harvester

mutantsan committed:
SXDEDPCXZIC-109 / write test for dcat_json harvester and helpers

mutantsan committed:
SXDEDPCXZIC-109 / append TestDelwpHarvester.test_import_stage test

joelwigley committed:
Merge pull request #31 from dpc-sdp/SXDEDPCXZIC-89_DATAVIC-517

SXDEDPCXZIC-89_DATAVIC-517 / prepare ODS description using markdownify

mutantsan committed:
SXDEDPCXZIC-129 / add temporal coverage for wms/wmf resources

alexmorev committed:
Merge branch 'SXDEDPCXZIC-31_DATAVIC-430' of https://github.com/agmorev/ckanext-datavic-harvester into agmorev-SXDEDPCXZIC-31_DATAVIC-430

alexmorev committed:
Merge pull request #1 from agmorev/agmorev-SXDEDPCXZIC-31_DATAVIC-430

SXDEDPCXZIC-31_DATAVIC-430 / resolve git conflicts

joelwigley committed:
Merge pull request #35 from mutantsan/SXDEDPCXZIC-129

SXDEDPCXZIC-129 / add temporal coverage for wms/wmf resources

alexmorev committed:
Merge branch 'uat' into SXDEDPCXZIC-31_DATAVIC-430
